presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,455 @@
+            return result;
+        }
+        return plan;
+    }
+
+    private boolean isEnabled(Session session)
+    {
+        return isOptimizePayloadJoins(session);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<JoinContext>
+    {
+        private final Session session;
+        Metadata metadata;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final VariableAllocator variableAllocator;
+
+        private Rewriter(Session session, Metadata metadata, FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.variableAllocator = requireNonNull(variableAllocator, "variableAllocator is null");
+        }
+
+        private Set<VariableReferenceExpression> extractPayloadColumns(List<VariableReferenceExpression> columns)
+        {
+            return columns.stream().filter(var -> var.getType() instanceof MapType || var.getType() instanceof ArrayType).collect(toImmutableSet());
+        }
+
+        @Override
+        public PlanNode visitPlan(PlanNode node, RewriteContext<JoinContext> context)
+        {
+            return context.defaultRewrite(node, context.get());
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context)
+        {
+            int numJoinKeys = context.get().getJoinKeys().size();
+            PlanNode leftNode = joinNode.getLeft();
+            PlanNode rightNode = joinNode.getRight();
+            boolean isTopJoin = numJoinKeys == 0;
+
+            List<VariableReferenceExpression> leftColumns = leftNode.getOutputVariables();
+            Set<VariableReferenceExpression> joinKeys = extractJoinKeys(joinNode.getFilter(), joinNode.getCriteria());
+
+            if (!needsRewrite(joinNode.getType(), leftColumns, joinKeys)) {
+                return context.defaultRewrite(joinNode, context.get());
+            }
+
+            Set<VariableReferenceExpression> leftJoinKeys = intersection(ImmutableSet.copyOf(leftColumns), joinKeys);
+
+            context.get().addKeys(leftJoinKeys);
+            context.get().incrementNumJoins();
+
+            PlanNode newLeftNode = context.rewrite(leftNode, context.get());
+            if (leftNode == newLeftNode) {
+                return context.defaultRewrite(joinNode, new JoinContext());
```

Comment:
It should be possible to skip all rewrite and return original plan node rather doing the default rewrite?

Reply:
there may be something else that needs rewriting further down the tree